src/Orleans.Transactions/State/ConfirmationWorker.cs

```diff
@@ -80,17 +80,21 @@ private async Task NotifyAll(Guid transactionId, DateTime timestamp, List<Partic
 
             // attempts to confirm all, will retry every ConfirmationRetryDelay until all succeed
             var ct = this.activationLifetime.OnDeactivating;
-            while (!ct.IsCancellationRequested && await HasPendingConfirmations(confirmations))
-            {
-                await this.timerManager.Delay(this.options.ConfirmationRetryDelay);
-            }
-
-            async Task<bool> HasPendingConfirmations(List<Confirmation> values)
+            var hasPendingConfirmations = true;
+            while (!ct.IsCancellationRequested && hasPendingConfirmations)
             {
                 using (this.activationLifetime.BlockDeactivation())
                 {
-                    var results = await Task.WhenAll(confirmations.Select(c => c.Confirmed()));
-                    return results.Any(b => !b);
+                    var confirmationResults = await Task.WhenAll(confirmations.Select(c => c.Confirmed()));
+                    hasPendingConfirmations = false;
+                    foreach (var confirmed in confirmationResults)
+                    {
+                        hasPendingConfirmations |= !confirmed;
+                    }
+
+                    if (!hasPendingConfirmations) break;
```

Comment:
if we're breaking out here, and hasPendingConfirmations can't change before loop starts again, why have hasPendingConfirmations as a loop condition?

Reply:
Good point